### ISSUE TITLE
The Apply To Selected Tokens checkbox overrides impersonation

### ISSUE THREAD

**Issue by Azhrei** [RPTools/maptool#2884]
**Describe the bug**
The Apply To Selected Token checkbox overrides the impersonation status of a token, so accidentally having it turned on leads to difficult to find errors.

**To Reproduce**
1. Open MapTool
2. Open the Impersonation panel
3. Drag any image to the map (PNG, JPEG, or .rptok)
4. Select the token and click the Impersonate button
5. Create a macro on that token.
6. Duplicate the macro to another name.
7. Change them so one has the **Apply to Selected Tokens** turned on and one does not.
8. Notice that there is no visual clue between the two macros.

**Expected behavior**
There are multiple approaches:

* ignore the checkbox when executing a macro on an impersonated token,
* turn the checkbox off by default when creating a macro on a token (or copying a macro from somewhere else, including importing),
* adding some kind of visual indicator so the macro writer knows the macro works differently, or
* some combination of the above.

**MapTool Info**
- Version: 1.9.3 (perhaps earlier)
- Install: New

**Desktop (please complete the following information):**
 - OS: macOS
 - Version 11.5.1 (fully updated)

**Additional context**
As it is, when a user creates macros _without_ the checkbox, they work correctly and always apply to the impersonated token.  When the checkbox is turned on, it overrides the impersonation, so the token must be selected to work properly.

**Comment by kwvanderlinde** [comment]
It's subtle, but there is already visual difference between the two macros: when the mouse is over a macro button with **Apply to Selected Tokens**, a white border is drawn around all selected tokens. This doesn't happen when **Apply to Selected Tokens** is turned off. (side note: the white border also shows up when hovering a macro in the "Selected" panel, regardless of **Apply to Selected Tokens**). All that said, it could be nice to make this difference obvious without requiring mouse hover. E.g., by adding an icon or some other indicator to the macro button itself.

Beyond that, this issue seems to boil down to improperly written macros (I'm making some assumptions here since the macros that triggered the creation of this issue haven't been provided). Here are my thoughts on best practices for writing macros:
1. Use `currentToken()` by default. It plays nicely with the `token:` roll option and `switchToken()`. Conceptually, this is the token that the macro is operating on, and is also the token that functions like `getName()` operate on when not passed a token ID.
2. If you really need the impersonated token, then you should use `getImpersonated(1)`. This always matches the user's concept of impersonated token. Keep in mind that there may not be an impersonated token.
3. If you really really need to get the macro context, then use `getImpersonated()`/`getImpersonated(0)`, or `getMacroContext()` if applicable. I find this rarely needed.
4. Consider using a couple of appropriately named UDF wrappers instead of calling `getImpersonated()` directly, to avoid confusion.

Using the right function for the job also addresses the first two suggestions mentioned under "expected behaviour":
- "ignore the checkbox when executing a macro on an impersonated token" A well-written macro won't really on `currentToken()` or `getImpersonated()` returning the impersonated token, whether or not **Apply to Selected Tokens** is selected. Instead, it would use `getImpersonated(1)` to get the impersonated token. Then, depending on whether the macro was designed for use with **Apply to Selected Tokens**, it will either use `currentToken()` or `getSelected()` to get the selected tokens. For example, a macro that deals damage independently to each selected token could reasonably be written with `getImpersonated(1)` to get the source of the damage, and then use `currentToken()` with **Apply to Selected Tokens** to find each target. The suggestion would break this use case.
- "turn the checkbox off by default..." Well, it is off by default. Only when copying it from elsewhere may the checkbox be on initially. And in that case, we should leave it as-is. Any macro written under the assumption of **Apply to Selected Tokens** being on would suddenly not function when copied to another location (e.g., the hypothetical targetting macro from the previous point).

**Comment by Azhrei** [comment]
Thanks for the reply; very useful. :)

> It's subtle, but there is already visual difference between the two macros: when the mouse is over a macro button with **Apply to Selected Tokens**, a white border is drawn around all selected tokens.

Thanks for pointing it out; I hadn’t noticed and neither had the user experiencing this issue (based on the Discord discussion).  I’ll check this out when I’m back in front of a computer.

> Here are my thoughts on best practices for writing macros:

Your best practices list is good, but users don’t know about them.  So they both need to be documented and (because users don’t read documentation) should be encouraged by the application.

If the Apply to Select Tokens were automatically turned off when a macro is created _on a token_, then the user would have no choice but to use one of these best practices, right?  And if they don’t know about these best practices, their macro still works as they expect.

(Maybe there’s a Preferences setting that causes this to be on by default for my setup, but I couldn’t find one.  Particularly since you state it was off by default later in your reply.  I may just be blind.)

> Using the right function for the job also addresses the first two suggestions mentioned under "expected behaviour":
> 
> * "ignore the checkbox when executing a macro on an impersonated token" A well-written macro […]

This was a novice user, so I think “well-written” is not on option. :)

Once I downloaded their campaign and looked at it myself, I knew what the problem was, but MT could’ve helped the user diagnose this themselves with a little help.

Visually distinguishing some macros compared to others would have at least triggered the question for the user, “why are these different?”, which maybe would’ve encouraged them to check the macro settings.  Then again, maybe not.  (A change to the tooltip help for the checkbox could at least specify something like, “this should probably be OFF when used with impersonated tokens”.)

> * "turn the checkbox off by default..." Well, it is off by default.

Hm.  It was ON when I tested it.  Which goes back to there being a Preferences setting that I missed.  I’ll look again when I’m back in front of my laptop.

**Comment by Phergus** [comment]
Applied to Selected is off by default.  There are two Preference settings for macros: whether Players can edit macros by default and enabling external access for macros.

If Apply to Selected is enabled on a macro it should apply whether there is an impersonated token or not.

As already noted MT very obviously highlights the selected tokens when a macro is going to act on them.

Not sure what you can do to help novice macro coders not shoot themselves in the foot when it comes to having Apply to Selected enabled but expecting it to work on an impersonated token.  Should MT do a check each time a macro is executed to see if there is an impersonated token and pop up a warning that Apply to Selected is enabled on the macro?  That would get annoying very quickly I think.

**Comment by kwvanderlinde** [comment]
> Once I downloaded their campaign and looked at it myself, I knew what the problem was ...

Out of curiousity, could you describe the situation further? Was this just a matter of using `getImpersonated()` and then being surprised it doesn't actually get the impersonated token? Because I'm basically operating under that assumption for the rest of my response :smile:  I've personally used `getImpersonated()` by mistake when writing my earlier macros, and I'm sure this is not an uncommon mistake.

> Your best practices list is good, but users don’t know about them. So they both need to be documented and (because users don’t read documentation) should be encouraged by the application.

I agree regarding the documentation and application encouragement. My main point was to make it clear later on that any novice-friendly solution should not punish macros that follow these practices.

That said, the wiki could be updated to have a lean best practices page for the differences between `currentToken()`, `getImpersonated()`, `getImpersonated(1)` and `getSelected()`. "Lean" meaning it won't overwhelm a novice with lots of information, but is a quick reference that will nudge novices in the right direction. The pages for each of the functions would then have a friendly suggestion to read the best practices page before using them. And `getImpersonated()` should really have a high-visibility warning label regarding its default behaviour, which would show up in the macro editor as well.

Going further, perhaps we can deprecate `getImpersonated()` in favour to two separate, unambiguously named functions. Though this is tricky considering the names `getMacroContext()` and `getImpersonated()` are probably the best names and they're both taken :frowning_face:

**Comment by Azhrei** [comment]
> Applied to Selected is off by default. There are two Preference settings for macros: whether Players can edit macros by default and enabling external access for macros.

Yep.  I tried it using an empty campaign and the checkbox is off.  I clearly screwed up that test. :(

> If Apply to Selected is enabled on a macro it should apply whether there is an impersonated token or not.

For a novice user, if they put a macro on a token, I would presume they expect it to apply to that token.  Since the AtST defaults to OFF, I'd say this is accomplished using the current code — no changes required.

If a novice user drags/imports a macro with AtST turned ON to a token, should anything change?  Maybe a warning that macros with AtST will override the impersonation?  Maybe they should put such macros in their Global panel?  I don't know.

> As already noted MT very obviously highlights the selected tokens when a macro is going to act on them.

Sometimes, yes.

The user in question had an issue where macros on impersonated tokens suddenly stopped working.  They didn't know why, but it happened to everyone (including the GM) multiple times throughout the session.

When AtST is ON:

* If there is no token selected, clicking the button does nothing.  There is no highlighting to warn the user, and there is no message in chat describing the issue (such as "AtST is ON but there are no selected tokens").  The highlighting could be on the button or the impersonated token, but the button is likely better.

* If there is a token selected, but the user doesn't own the token (meaning STO must be ON since turning it off implies players own all tokens?), clicking the button produces an error message in chat ("Only _Token_'s owners can impersonate or run macros on it.").  This is better, but why not provide a visual clue when hovering over the button itself?

    * How does the server option, "Allow Players to Impersonate any token" affect this?  I don't know if the code treats this as making players implicit owners of all tokens, or if it's a special flag implemented only in the Impersonation panel or somewhere else.

When AtST is OFF:

* If there is or is not a token selected, clicking the button works and applies the macro against the impersonated token.  This is fine and is what a user would expect, although adding the glow to the impersonated token would be a nice enhancement and would be in line with the way it works when AtST _is_ turned on.

> Not sure what you can do to help novice macro coders not shoot themselves in the foot when it comes to having Apply to Selected enabled but expecting it to work on an impersonated token. Should MT do a check each time a macro is executed to see if there is an impersonated token and pop up a warning that Apply to Selected is enabled on the macro? That would get annoying very quickly I think.

Definitely not!  But a check could be made when the AtST checkbox is turned on, when a macro is dragged to a token and has the checkbox on, or when a macro is imported that has the checkbox on.  Those are all discrete events initiated by the user.  (Maybe importing an entire macro set should give the option of reseting that flag for all macros in the set, just like importing a map offers to clear any existing FOW.)

(This got long, so I'm going to respond kwvanderlinde separately.)

**Comment by Azhrei** [comment]
> > Once I downloaded their campaign and looked at it myself, I knew what the problem was ...
> 
> Out of curiousity, could you describe the situation further? Was this just a matter of using `getImpersonated()` and then being surprised it doesn't actually get the impersonated token? Because I'm basically operating under that assumption for the rest of my response 😄 I've personally used `getImpersonated()` by mistake when writing my earlier macros, and I'm sure this is not an uncommon mistake.

No.  The macros I looked at were all very template-ish and only accessed properties on the token (without using get/set property functions).  I saw no use of any functions in regards to token context:  no use of `switchToken()`, `getImpersonated()`, `getSelected()`, or similar.

This is a novice user just pounding on a keyboard until something worked, then being confused when it suddenly stopped working...

Now that I know the AtST defaults to off, it's clear that the user specifically turned it on.  Which is why I think the simplest solution to prevent this in the future is just to warn users when they turn it on (whether interactively, via drag/drop, or importing a macro).

It's obviously fine to skip warnings for Campaign panel, Global panel, and GM panel macros.  I don't think there would ever be any confusion that there is an implied "impersonated" token for those.

> That said, the wiki could be updated to have a lean best practices page for the differences between `currentToken()`, `getImpersonated()`, `getImpersonated(1)` and `getSelected()`.

I don't have enough experience with those functions to have an opinion on the best way to handle any documentation updates.  I'm just relaying the information provided in the Discord chat and in my discovery of the UX confusion. :)

**Comment by Phergus** [comment]
> For a novice user, if they put a macro on a token, I would presume they expect it to apply to that token.

And if the macro is named "Heal Other" or "Range to Target"?

> If a novice user drags/imports a macro with AtST turned ON to a token, should anything change? Maybe a warning that macros with AtST will override the impersonation?

What if there are no impersonated tokens should there still be a warning?  Perhaps only if the macro is dropped on the Impersonated panel?  Or perhaps just a general warning that the macro has AtST enabled?  Of course the first thing that will happen is the warning will get turned off and everyone will go back to not remembering how current token, selected token and impersonated all work.

> Maybe they should put such macros in their Global panel? I don't know.

If the macro writer has set a macro to AiST it seems like it's probably a macro that was meant to stay on the Campaign panel.

> When AtST is ON:
> 
>     * If there is no token selected, clicking the button does nothing.  There is no highlighting to warn the user, and there is no message in chat describing the issue (such as "AtST is ON but there are no selected tokens").

I could have sworn there used to be either an error dialog or a chat warning for when a macro had AtST enabled but no tokens were selected but I guess not.  Adding a highlight to the button in a different color when the macro has AtST but no tokens are selected seems like a good idea.

>     * If there is a token selected, but the user doesn't own the token (meaning STO must be ON since turning it off implies players own all tokens?), clicking the button produces an error message in chat ("Only _Token_'s owners can impersonate or run macros on it.").  This is better, but why not provide a visual clue when hovering over the button itself?

Should that be happening for trusted macros?  I wonder if something changed.

>     * How does the server option, "Allow Players to Impersonate any token" affect this?  I don't know if the code treats this as making players implicit owners of all tokens, or if it's a special flag implemented only in the Impersonation panel or somewhere else.

No idea.

> When AtST is OFF:
> 
>     * If there is or is not a token selected, clicking the button works and applies the macro against the impersonated token.  This is fine and is what a user would expect, although adding the glow to the impersonated token would be a nice enhancement and would be in line with the way it works when AtST _is_ turned on.

Highlighting the tokens to be affected by any macro seems like a good idea.

> > Should MT do a check each time a macro is executed to see if there is an impersonated token and pop up a warning that Apply to Selected is enabled on the macro?
> 
> Definitely not! But a check could be made when the AtST checkbox is turned on, when a macro is dragged to a token and has the checkbox on, or when a macro is imported that has the checkbox on. Those are all discrete events initiated by the user. (Maybe importing an entire macro set should give the option of reseting that flag for all macros in the set, just like importing a map offers to clear any existing FOW.)

As long as every one of those warnings can be disabled.

**Comment by Azhrei** [comment]
> > For a novice user, if they put a macro on a token, I would presume they expect it to apply to that token.
> 
> And if the macro is named "Heal Other" or "Range to Target"?

You and I must have different definitions of "novice user".  I'm thinking about users who have (perhaps) never visited the wiki and have learned everything they know by copy/pasting from other people's macros.

> > If a novice user drags/imports a macro with AtST turned ON to a token, should anything change? Maybe a warning that macros with AtST will override the impersonation?
> 
> What if there are no impersonated tokens should there still be a warning? Perhaps only if the macro is dropped on the Impersonated panel? Or perhaps just a general warning that the macro has AtST enabled? Of course the first thing that will happen is the warning will get turned off and everyone will go back to not remembering how current token, selected token and impersonated all work.

All good points.

IMO, macros on tokens should 99% of the time be meant to apply to that token.  Your example of "Heal Other" or similar targeting macros are the exception, but I'm not convinced it would have AtST turned on anyway.  (If it did, it would prompt for and apply healing to all selected tokens one a time but never update the spell slot/mana points/whatever on the token that was taking that action.  This is clearly going to require the "novice" user to graduate to the "apprentice" level and learn about `getSelected()` and similar functions.)

> > Maybe they should put such macros in their Global panel? I don't know.
> 
> If the macro writer has set a macro to AiST it seems like it's probably a macro that was meant to stay on the Campaign panel.

Well, users wouldn't have that option, hence "Global", but you're right, the Campaign panel is the place to keep it so that other players and the GM could use it as well.

> I could have sworn there used to be either an error dialog or a chat warning for when a macro had AtST enabled but no tokens were selected but I guess not. Adding a highlight to the button in a different color when the macro has AtST but no tokens are selected seems like a good idea.
> 
> > ```
> > * If there is a token selected, but the user doesn't own the token (meaning STO must be ON since turning it off implies players own all tokens?), clicking the button produces an error message in chat ("Only _Token_'s owners can impersonate or run macros on it.").  This is better, but why not provide a visual clue when hovering over the button itself?
> > ```
> 
> Should that be happening for trusted macros? I wonder if something changed.

Hm, good point.  I hadn't considered trusted macros.  I suppose any button highlighting could take that into account since the trusted state is essentially static once the macro is created (meaning that the state doesn't change unless the user changes something).

I'm beginning to think this is going to result in a spreadsheet like the one Wolph made way back when for vision, light, and token ownership...

> > ```
> > * How does the server option, "Allow Players to Impersonate any token" affect this?  I don't know if the code treats this as making players implicit owners of all tokens, or if it's a special flag implemented only in the Impersonation panel or somewhere else.
> > ```
> 
> No idea.
> 
> > When AtST is OFF:
> > ```
> > * If there is or is not a token selected, clicking the button works and applies the macro against the impersonated token.  This is fine and is what a user would expect, although adding the glow to the impersonated token would be a nice enhancement and would be in line with the way it works when AtST _is_ turned on.
> > ```
> 
> Highlighting the tokens to be affected by any macro seems like a good idea.

Agreed.

It gives me an idea for another FREQ...

> > Definitely not! But a check could be made when the AtST checkbox is turned on, when a macro is dragged to a token and has the checkbox on, or when a macro is imported that has the checkbox on. Those are all discrete events initiated by the user. (Maybe importing an entire macro set should give the option of reseting that flag for all macros in the set, just like importing a map offers to clear any existing FOW.)
> 
> As long as every one of those warnings can be disabled.

If you want; I'm ambivalent about it.  The existing warning I referenced earlier that appears in chat cannot be turned off.  The prompt for resetting the FOW on map import cannot be turned off.  There are plenty of such warnings that can't be turned off, I presume because they don't happen frequently enough for it to be a hassle responding to them.

I can't imagine the prompt discussed herein would happen enough to be a hassle either.  How often is a macro with AtST going to be copied to a token anyway?  If anything, the copies are probably done mostly in the opposite direction — where a player creates a macro for their token and the GM decides it warrants use by everyone.

**Comment by Phergus** [comment]
> If there is a token selected, but the user doesn't own the token (meaning STO must be ON since turning it off implies players own all tokens?), clicking the button produces an error message in chat ("Only Token's owners can impersonate or run macros on it."). 

This is a bug.  Was broken sometime between 1.7.0 and 1.8.2.

**Comment by Azhrei** [comment]
> > If there is a token selected, but the user doesn't own the token (meaning STO must be ON since turning it off implies players own all tokens?), clicking the button produces an error message in chat ("Only Token's owners can impersonate or run macros on it.").
> 
> This is a bug. Was broken sometime between 1.7.0 and 1.8.2.

I'm not sure that's a bug — I *like* that functionality!  The message even makes sense and describes exactly what the problem is!

Do you have a particular reason why you think it's inappropriate...?

**Comment by Phergus** [comment]
Because it defeats the point of trusted macros.
